src/App.tsx

```diff
@@ -1,26 +1,216 @@
+  const [error, setError] = useState<string | null>(null);
+  const [status, setStatus] = useState<TodoStatus>(TodoStatus.All);
+  const [selectedTodo, setSelectedTodo] = useState<Todo | null>(null);
+  const [processingIds, setProcessingIds] = useState<Todo['id'][]>([]);
+  const [isAdding, setIsAdding] = useState(false);
+
+  useEffect(() => {
+    setError(null);
+    getTodos()
+      .then(setTodos)
+      .catch(() => setError(ERROR_MESSAGES.LOAD_TODOS));
+  }, []);
+
+  useEffect(() => {
+    if (error) {
+      const timer = setTimeout(() => setError(null), 3000);
+
+      return () => clearTimeout(timer);
+    }
+  }, [error]);
+
+  const filteredTodos = useMemo(() => {
+    const list = [...todos];
+
+    if (tempTodo) {
+      list.push(tempTodo);
+    }
+
+    return list.filter(todo => {
+      switch (status) {
+        case 'active':
+          return !todo.completed;
+        case 'completed':
+          return todo.completed;
+        default:
+          return true;
+      }
+    });
+  }, [todos, tempTodo, status]);
+
+  useEffect(() => {
+    if (!isAdding && !selectedTodo) {
+      const input = document.querySelector<HTMLInputElement>(
+        'input.todoapp__new-todo',
+      );
+
+      input?.focus();
+    }
+  }, [isAdding, todos.length, selectedTodo]);
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
+  const handleAddTodo = (event: React.FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+    const form = event.currentTarget;
+    const input = form.elements.namedItem('NewTodoField') as HTMLInputElement;
+    const title = input.value.trim();
+
+    if (!title) {
+      setError(ERROR_MESSAGES.EMPTY_TITLE);
+
+      return;
+    }
+
+    const tempId = Date.now();
+    const newTemp: Todo = {
+      id: tempId,
+      userId: USER_ID,
+      title,
+      completed: false,
+    };
+
+    setTempTodo(newTemp);
+    setProcessingIds(prev => [...prev, tempId]);
+    setIsAdding(true);
+    setError(null);
+
+    addTodo(title)
+      .then(newTodo => {
+        setTodos(current => [...current, newTodo]);
+        setTempTodo(null);
+        form.reset();
+      })
+      .catch(() => {
+        setTempTodo(null);
+        setError(ERROR_MESSAGES.ADD_TODO);
+      })
+      .finally(() => {
+        setProcessingIds(prev => prev.filter(id => id !== tempId));
+        setIsAdding(false);
+      });
+  };
+
+  const handleDeleteTodo = (todoId: number) => {
+    setError(null);
+    setProcessingIds(prev => [...prev, todoId]);
+
+    deleteTodo(todoId)
+      .then(() => {
+        setTodos(currentTodos => currentTodos.filter(t => t.id !== todoId));
+        if (selectedTodo?.id === todoId) {
+          setSelectedTodo(null);
+        }
+      })
+      .catch(() => setError(ERROR_MESSAGES.DELETE_TODO))
+      .finally(() =>
+        setProcessingIds(prev => prev.filter(id => id !== todoId)),
+      );
+  };
+
+  const handleUpdateTodo = (todoId: number, data: Partial<Todo>) => {
```

Comment:
ideally :
```suggestion
  const handleUpdateTodo = (todoId: number, data: Partial<Omit<Todo, 'id'>>) => {
```

to indicate which id is correct

or even some TS kung fu
```suggestion
  const handleUpdateTodo = (todo: { id: Todo["id"] } & Partial<Omit<Todo, "id">>) => {
  const {id, ...todoData} = todo;
  // do your things
  
```
in that sample id is required, while all other fields are optional
IMO - it looks even more ellegant